### PR TITLE
Make historical Airtable tables

### LINKED ISSUE #1868: GTFS schedule refactor -- warehouse: Make Airtable tables historical
To ensure we can make the most of our GTFS data, we need type 2 Airtable data that contains prior data with effective dates (instead of only surfacing the current Airtable data, as we do now).

For this we should:
Change `gtfs_datasets`, `gtfs_service_data`, `services`, and `organizations` to: 

* Stop filtering to only one extract per day in staging
* Stop getting latest-only in mart 
* Start using a versioned key (instead of just using the Airtable record ID) -- surrogate key from record ID + extract timestamp. 
   * In the case of `gtfs datasets` in particular, we may want to use base64_url + timestamp, although this will introduce some complexity. We will need to handle historical cases where duplicate or null URLs were present.
* Update the bridge & synced table handling accordingly; this will be challenging because not all Airtable extracts will have the same timestamps

### LINKED ISSUE #2102: Add a link to the data documentation in warehouse table docs
There is a Google doc that has data documentation about Airtable (not linking here until I can double check that it has appropriate sharing permissions). That Google doc should be linked in the warehouse definitions of the tables described in it (organizations, services, GTFS service data, GTFS datasets) so that people can access it from dbt documentation. 

(This is a follow up from the Airtable Recurring meeting 12/21/22)

### PR DESCRIPTION
**Pull request by lauriemerrell**
# Description

_After this merges, we need to run a full refresh to change all the keying everywhere_

This PR:
- Makes Airtable dimensions for organizations, services, GTFS datasets, and GTFS service data historical instead of current-only
- Adds fake versioning columns to the other Airtable data, to prepare for eventual proper historicization
- Propagates updated foreign keys associated with the above to GTFS and GTFS quality data 
   - After PR merge, this will require a full refresh in prod to actually backfill these versioned keys. 
   - **I think that we should do #2091 before doing the full refresh for this PR.**
- Removes tests on staging Airtable tables in the interest of cost efficiency (especially since this adds intermediate materialized tables that can be tested instead)

As part of this I am changing/refactoring a bunch of the work I had done in #2087 in terms of intermediate materializations of things.

This PR does NOT add columns requested in #2091 -- will do that as an immediate follow up, I just didn't want to touch Airflow in this (already enormous) PR. 

<details>
<summary> Completed to-do items </summary>

- [x] Dimension tables (working macro & apply to all tables)
- [x] Rename date column in staging tables 
- [x] Overlapping ranges tests for dimensions (including fake ones)
- [x] Deal with versioned foreign keys:
   - [x] `dim_contract_attachments` needs proper foreign keys (will require making intermediate tables)
   - [x] `dim_service_components` needs proper foreign keys  (will require making intermediate tables)
   - [x] `dim_provider_gtfs_data` and its upstream quartet table
   - [x] Check for other foreign keys, for example to/from GTFS data 
- [x] Add fake historical columns to the remaining Airtable tables 
- [x] Bridge tables that handle historicity
   - [x] handle `_is_current` 
- [x] Update docs per #2102 
- [x] Update work in #2087 to use the new dimensions 
- [x] Make a ticket to handle the wrongly-entered `gtfs service data` records -- #2174
- [x] Figure out urls to datasets 
- [x] Handle all the YAML
- [x] Add NTD ID back to dim provider GTFS data
- [x] Handle Torrance missing schedule -- decided to defer 

Questions to discuss:
* How to handle the fact that we are currently only keeping a very small subset of columns... Should versioning be based only on columns that we are surfacing in the final dimension table, or all columns? How urgent is it to get all the actual columns into the staging table? (Currently, staging is a subset of external and mart is a subset of staging.) 
   * Verdict: Versioning should only be based on columns that are available in the mart.  
* Should we remove lookup columns from type 2 logic? There are a lot on GTFS service data.  
   * Verdict: Yes, remove lookup columns. 

</details>

Resolves #2102 , resolves #1868

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Documentation
- [ ] agencies.yml

## How has this been tested?

I have tested most portions of the warehouse locally

### CODE REVIEW

File: warehouse/models/intermediate/transit_database/int_transit_database__urls_to_gtfs_datasets.sql [hunk 1]
```diff
@@ -1,21 +1,44 @@
...
+        gtfs_datasets.original_record_id,
+        gtfs_datasets.key AS gtfs_dataset_key,
+        CASE
+            WHEN gtfs_datasets._valid_from = appearance_duration.first_app THEN CAST('1900-01-01' AS TIMESTAMP)
+            ELSE gtfs_datasets._valid_from
+        END AS _valid_from,
+        gtfs_datasets._valid_to
+    FROM gtfs_datasets
+    LEFT JOIN appearance_duration
+        USING (base64_url)
```
Comment (atvaccaro) [inline warehouse/models/intermediate/transit_database/int_transit_database__urls_to_gtfs_datasets.sql:32]: I'm surprised sqlfluff didn't alert on this

File: warehouse/models/mart/transit_database/_mart_transit_database.yml [hunk 2]
```diff
@@ -657,3 +664,49 @@ models:
...
+      - name: gtfs_dataset_key
+        tests:
+          - relationships:
+             to: ref('dim_gtfs_datasets')
+             field: key
+      - name: schedule_to_use_for_rt_validation_gtfs_dataset_key
+        tests:
+          - relationships:
+             to: ref('dim_gtfs_datasets')
+             field: key
```
Comment (atvaccaro) [inline warehouse/models/mart/transit_database/_mart_transit_database.yml:712]: nit: would it make sense to use https://github.com/dbt-labs/dbt-utils#relationships_where-source to test gtfs_dataset_key only ever points to RT and schedule_to_use_for_rt_validation_gtfs_dataset_key only ever points to Schedule?

**Review by atvaccaro — APPROVED** [review]
lgtm! few small requests/questions but nothing blocking

int_gtfs_quality__daily_assessment_candidate_entities and dim_provider_gtfs_data are definitely complex so thanks for documenting them thoroughly

I couldn't put this where I wanted but for int_gtfs_quality__daily_assessment_candidate_entities should the key also contain the date, rather than unique combinations test on key and date?

File: warehouse/models/staging/transit_database/stg_transit_database__gtfs_service_data.sql [hunk 3]
```diff
@@ -22,27 +21,16 @@ stg_transit_database__gtfs_service_data AS (
...
-        dt AS calitp_extracted_at
+        dt
     FROM once_daily_gtfs_service_data
     LEFT JOIN UNNEST(once_daily_gtfs_service_data.services) as unnested_services
     LEFT JOIN UNNEST(once_daily_gtfs_service_data.gtfs_dataset) as unnested_gtfs_dataset
+    -- TODO: actually handle this -- there are records that were not 1:1
+    -- either one service was entered with multiple datasets in a single record
+    -- or vice versa
+    -- we should generate a new key and find a way to keep both
+    QUALIFY ROW_NUMBER() OVER (PARTITION BY id, ts ORDER BY service_key) = 1
```
[the remark this answers is not in the document]
Comment (lauriemerrell) [reply warehouse/models/staging/transit_database/stg_transit_database__gtfs_service_data.sql:33]: It was never supposed to be used that way (table always intended to be 1:1), but yeah. I think there are less than 5 affected records ever.

[the remark this answers is not in the document]
Comment (lauriemerrell) [reply warehouse/models/staging/transit_database/stg_transit_database__gtfs_service_data.sql:33]: I am clarifying the comment

File: warehouse/macros/transit_database_make_historical_dimension.sql [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+            {% endfor %}
+            {% endif %}
+        )
+    {% endif %}
+    {% if array_cols %},
+    {% for col in array_cols %}
+        ARRAY_TO_STRING({{ col }}, '--') AS {{ col }} {% if not loop.last %},{% endif %}
+    {% endfor %}
+    {% endif %}
+    FROM {{ ref(once_daily_staging_table) }}
```
[the remark this answers is not in the document]
Comment (lauriemerrell) [reply warehouse/macros/transit_database_make_historical_dimension.sql:29]: added a todo comment

File: warehouse/macros/transit_database_make_historical_dimension.sql [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+-- following: https://dba.stackexchange.com/questions/210907/determine-consecutive-occurrences-of-values
+first_instances AS (
+    SELECT
+        hashed.dt,
+        key,
+        latest_extract,
+        (DENSE_RANK() OVER (ORDER BY latest_extract DESC)) = 1 AS in_latest,
+        next_dt,
+        (LAG(content_hash) OVER (PARTITION BY key ORDER BY hashed.dt) != content_hash)
+            OR (LAG(content_hash) OVER (PARTITION BY key ORDER BY hashed.dt) IS NULL) AS is_first
```
[the remark this answers is not in the document]
Comment (lauriemerrell) [reply warehouse/macros/transit_database_make_historical_dimension.sql:63]: yes

File: warehouse/models/intermediate/transit_database/int_transit_database__urls_to_gtfs_datasets.sql (hunk 1 above)
Comment (lauriemerrell) [reply warehouse/models/intermediate/transit_database/int_transit_database__urls_to_gtfs_datasets.sql:32]: why?

Comment (atvaccaro) [reply warehouse/models/intermediate/transit_database/int_transit_database__urls_to_gtfs_datasets.sql:32]: oh I forgot we'd disabled https://docs.sqlfluff.com/en/stable/rules.html#sqlfluff.rules.Rule_L032

File: warehouse/models/mart/transit_database/dim_service_components.sql [hunk 6]
```diff
@@ -1,62 +1,27 @@
 {{ config(materialized='table') }}
 
-WITH latest_service_components AS (
+WITH int_transit_database__service_components AS (
     SELECT *
-    FROM {{ ref('int_transit_database__service_components_unnested') }}
+    FROM {{ ref('int_transit_database__service_components_dim') }}
 ),
```
[the remark this answers is not in the document]
Comment (atvaccaro) [reply warehouse/models/mart/transit_database/dim_service_components.sql:6]: that works for me

**Pull request merged**